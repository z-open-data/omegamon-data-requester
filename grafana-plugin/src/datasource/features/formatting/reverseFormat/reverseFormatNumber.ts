import { convertToNumberOrThrow } from './convertToNumberOrThrow';

function isValidDec(val: string): boolean {
  return /^\s*-?\d+(\.\d+)?\s*$/.test(val);
}

function isValidHex(val: string): boolean {
  return /^\s*0x[0-9a-f]+\s*$/i.test(val);
}

/**
 * @param value - hex number (
 *   without minus `-` sign, cause the `-` sign has been
 *   encoded into number itself using 2s complement method
 * )
 */
function reverseFormatHexNumber(value: string): number {
  const withoutSign = convertToNumberOrThrow(value);
  const withRestoredSign = withoutSign | 0;
  return withRestoredSign;
}

function reverseScale(value: number | string, scaleFactor?: number): number {
  const asNumber = convertToNumberOrThrow(value);
  if (scaleFactor == null) {
    return asNumber;
  }
  return asNumber * 10 ** scaleFactor;
}

/**
 * Converts number from human-readable format back to raw ITM format.
 * @param value - e.g. `"0XA"` or `"A"`, or `10`
 * @param scaleFactor - scaleFactor from column metadata
 * @returns decimal number, e.g. `10`
 *
 * Note:
 * the returned number may or may not be exactly in the same form as in ITM,
 *
 * E.g.
 * in ITM value for COLUMN_X may be `"0xA"`.
 * To successfully match it in WHERE clause, you can write any of the following:
 * 1. WHERE COLUMN_X = 0xA
 * 2. WHERE COLUMN_X = 0xa
 * 1. WHERE COLUMN_X = 10
 */
export function reverseFormatNumber(value: number | string, scaleFactor: number | undefined): number {
  if (typeof value === 'number' || isValidDec(value)) {
    return reverseScale(value, scaleFactor);
  }
  if (isValidHex(value)) {
    return reverseFormatHexNumber(value);
  }
  throw new Error(`reverseFormatNumber: Failed to convert string "${value}" to number`);
}
