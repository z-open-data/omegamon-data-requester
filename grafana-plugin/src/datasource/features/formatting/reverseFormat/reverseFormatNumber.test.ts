import {
  negativeNumbersAsHex2sComplements,
  nonNegativeNumbersAsHex,
  simpleNumbers,
  hexNumbers,
  stringifiedDecimalNumbers,
} from './mocks';
import { reverseFormatNumber } from './reverseFormatNumber';

describe(reverseFormatNumber.name, () => {
  describe('if hex number is 4 byte length (0xHHHHHHHH) with left-most bit as `1`, then this hex is converted into negative decimal', () => {
    negativeNumbersAsHex2sComplements.forEach((val) => {
      it(val, () => {
        const decimalNumber = reverseFormatNumber(val, undefined);
        expect(decimalNumber).toBeLessThan(0);
      });
    });
  });

  describe('other hex numbers are converted to non-negative decimal', () => {
    nonNegativeNumbersAsHex.forEach((val) => {
      it(val, () => {
        const decimalNumber = reverseFormatNumber(val, undefined);
        expect(decimalNumber).toBeGreaterThanOrEqual(0);
        expect(decimalNumber).toEqual(parseInt(val, 16));
      });
    });
  });

  describe('returns same number, when passed simple JavaScript number without scaleFactor', () => {
    simpleNumbers.forEach((val) => {
      it(`${val}`, () => {
        expect(reverseFormatNumber(val, undefined)).toEqual(val);
      });
    });
  });

  describe('correctly handles scaleFactor', () => {
    simpleNumbers.forEach((val) => {
      it(`${val}`, () => {
        expect(reverseFormatNumber(val, 3)).toEqual(val * 1000);
      });
    });
  });

  describe('throws error, if supplied with string that does not hold valid number', () => {
    const unconvertableStrings = ['', '   ', 'iCannotBeConvertedToNumber', '420bla', 'bla420', '0xABCDEFG', '0XNN'];
    unconvertableStrings.forEach((val) => {
      it(`"${val}"`, () => {
        expect(() => reverseFormatNumber(val, undefined)).toThrowError();
      });
    });
  });

  describe('hex', () => {
    hexNumbers.forEach((val) => {
      it(`"${val}"`, () => {
        const decimalNumber = reverseFormatNumber(val, undefined);
        expect(decimalNumber).toMatchSnapshot();
      });
    });
  });

  describe('dec', () => {
    stringifiedDecimalNumbers.forEach((val) => {
      it(`"${val}"`, () => {
        const decimalNumber = reverseFormatNumber(val, undefined);
        expect(decimalNumber).toMatchSnapshot();
      });
    });
  });
});
