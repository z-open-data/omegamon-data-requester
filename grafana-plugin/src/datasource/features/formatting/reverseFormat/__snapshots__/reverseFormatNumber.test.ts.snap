// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reverseFormatNumber dec "  -0.123  " 1`] = `-0.123`;

exports[`reverseFormatNumber dec "  123  " 1`] = `123`;

exports[`reverseFormatNumber dec "-0.723" 1`] = `-0.723`;

exports[`reverseFormatNumber dec "-1" 1`] = `-1`;

exports[`reverseFormatNumber dec "-123" 1`] = `-123`;

exports[`reverseFormatNumber dec "0" 1`] = `0`;

exports[`reverseFormatNumber dec "0.00234" 1`] = `0.00234`;

exports[`reverseFormatNumber dec "0.1" 1`] = `0.1`;

exports[`reverseFormatNumber dec "000123" 1`] = `123`;

exports[`reverseFormatNumber dec "1" 1`] = `1`;

exports[`reverseFormatNumber dec "10" 1`] = `10`;

exports[`reverseFormatNumber dec "123" 1`] = `123`;

exports[`reverseFormatNumber dec "14834921" 1`] = `14834921`;

exports[`reverseFormatNumber hex "0X0" 1`] = `0`;

exports[`reverseFormatNumber hex "0X000ffffffff" 1`] = `-1`;

exports[`reverseFormatNumber hex "0X00abc" 1`] = `2748`;

exports[`reverseFormatNumber hex "0X0A" 1`] = `10`;

exports[`reverseFormatNumber hex "0x0" 1`] = `0`;

exports[`reverseFormatNumber hex "0x000C" 1`] = `12`;

exports[`reverseFormatNumber hex "0x1" 1`] = `1`;

exports[`reverseFormatNumber hex "0x8A3A1234" 1`] = `-1975905740`;

exports[`reverseFormatNumber hex "0x1234567" 1`] = `19088743`;

exports[`reverseFormatNumber hex "0x80000000" 1`] = `-2147483648`;

exports[`reverseFormatNumber hex "0xA" 1`] = `10`;

exports[`reverseFormatNumber hex "0xA1234567" 1`] = `-1591523993`;

exports[`reverseFormatNumber hex "0xABCDEF" 1`] = `11259375`;

exports[`reverseFormatNumber hex "0xFFFFBA98" 1`] = `-17768`;

exports[`reverseFormatNumber hex "0xFFFFFFFF" 1`] = `-1`;
